#! /bin/sh -e

# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: Make glibc-2.3.5 compile with gcc-4.0 on arm for inline
#		   functions of dl-machine.h.
# DP: Related bugs: 
# DP: Dpatch author: GOTO Masanori <gotom@debian.org>
# DP: Patch author: Jakub Jelinek <jakub@redhat.com>
#		    GOTO Masanori <gotom@debian.org>
# DP: Upstream status: In CVS / Debian-Specific
# DP: Status Details: 
# DP: Date: 2005-07-17, updated 2005-07-20

PATCHLEVEL=0

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p$PATCHLEVEL < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p$PATCHLEVEL < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.
2005-07-20  GOTO Masanori  <gotom@debian.org>

	* sysdeps/sparc/sparc64/dl-machine.h: Change static/auto depended on
	the definition of RESOLVE.

2005-07-17  GOTO Masanori  <gotom@debian.org>

	* sysdeps/sparc/sparc32/dl-machine.h: Use auto inline.
	* sysdeps/sparc/sparc64/dl-machine.h: Likewise.

--- sysdeps/sparc/sparc32/dl-machine.h.org	2005-07-15 06:14:11.000000000 +0900
+++ sysdeps/sparc/sparc32/dl-machine.h	2005-07-15 06:59:13.000000000 +0900
@@ -411,7 +411,14 @@ elf_machine_plt_value (struct link_map *
 /* Perform the relocation specified by RELOC and SYM (which is fully resolved).
    MAP is the object containing the reloc.  */
 
-static inline void
+#if __GNUC__ >= 4
+  auto inline void
+#else
+  static inline void
+#endif
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2)
+  __attribute ((always_inline))
+#endif
 elf_machine_rela (struct link_map *map, const Elf32_Rela *reloc,
 		  const Elf32_Sym *sym, const struct r_found_version *version,
 		  void *const reloc_addr_arg)
@@ -593,7 +600,14 @@ elf_machine_rela (struct link_map *map, 
     }
 }
 
-static inline void
+#if __GNUC__ >= 4 
+  auto inline void 
+#else 
+  static inline void 
+#endif 
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2) 
+  __attribute ((always_inline)) 
+#endif 
 elf_machine_rela_relative (Elf32_Addr l_addr, const Elf32_Rela *reloc,
 			   void *const reloc_addr_arg)
 {
@@ -601,7 +615,14 @@ elf_machine_rela_relative (Elf32_Addr l_
   *reloc_addr += l_addr + reloc->r_addend;
 }
 
-static inline void
+#if __GNUC__ >= 4 
+  auto inline void 
+#else 
+  static inline void 
+#endif 
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2) 
+  __attribute ((always_inline)) 
+#endif 
 elf_machine_lazy_rel (struct link_map *map,
 		      Elf32_Addr l_addr, const Elf32_Rela *reloc)
 {
--- sysdeps/sparc/sparc64/dl-machine.h.org	2004-03-05 19:27:55.000000000 +0900
+++ sysdeps/sparc/sparc64/dl-machine.h	2005-07-20 11:21:06.000000000 +0900
@@ -34,7 +34,18 @@
 #define ELF64_R_TYPE_DATA(info) ((info) >> 8)
 
 /* Return nonzero iff ELF header is compatible with the running host.  */
-static inline int
+#if __GNUC__ >= 4
+# ifdef RESOLVE
+    auto inline int
+# else
+    static inline int
+# endif
+#else
+  static inline int
+#endif
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2)
+  __attribute ((always_inline))
+#endif
 elf_machine_matches_host (const Elf64_Ehdr *ehdr)
 {
   return ehdr->e_machine == EM_SPARCV9;
@@ -55,7 +66,18 @@
 /* Return the link-time address of _DYNAMIC.  Conveniently, this is the
    first element of the GOT.  This must be inlined in a function which
    uses global data.  */
-static inline Elf64_Addr
+#if __GNUC__ >= 4
+# ifdef RESOLVE
+    auto inline Elf64_Addr
+# else
+    static inline Elf64_Addr
+# endif
+#else
+  static inline Elf64_Addr
+#endif
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2)
+  __attribute ((always_inline))
+#endif
 elf_machine_dynamic (void)
 {
   register Elf64_Addr *elf_pic_register __asm__("%l7");
@@ -66,7 +88,18 @@
 }
 
 /* Return the run-time load address of the shared object.  */
-static inline Elf64_Addr
+#if __GNUC__ >= 4
+# ifdef RESOLVE
+    auto inline Elf64_Addr
+# else
+    static inline Elf64_Addr
+# endif
+#else
+  static inline Elf64_Addr
+#endif
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2)
+  __attribute ((always_inline))
+#endif
 elf_machine_load_address (void)
 {
   register Elf32_Addr *pc __asm ("%o7");
@@ -88,7 +121,18 @@
 
 /* We have 4 cases to handle.  And we code different code sequences
    for each one.  I love V9 code models...  */
-static inline void
+#if __GNUC__ >= 4
+# ifdef RESOLVE
+    auto inline void 
+# else
+    static inline void
+# endif
+#else
+  static inline void
+#endif
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2)
+  __attribute ((always_inline))
+#endif
 sparc64_fixup_plt (struct link_map *map, const Elf64_Rela *reloc,
 		   Elf64_Addr *reloc_addr, Elf64_Addr value,
 		   Elf64_Addr high, int t)
@@ -212,7 +256,18 @@
     }
 }
 
-static inline Elf64_Addr
+#if __GNUC__ >= 4
+# ifdef RESOLVE
+    auto inline Elf64_Addr
+# else
+    static inline Elf64_Addr
+# endif
+#else
+  static inline Elf64_Addr
+#endif
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2)
+  __attribute ((always_inline))
+#endif
 elf_machine_fixup_plt (struct link_map *map, lookup_t t,
 		       const Elf64_Rela *reloc,
 		       Elf64_Addr *reloc_addr, Elf64_Addr value)
@@ -223,7 +278,18 @@
 }
 
 /* Return the final value of a plt relocation.  */
-static inline Elf64_Addr
+#if __GNUC__ >= 4
+# ifdef RESOLVE
+    auto inline Elf64_Addr
+# else
+    static inline Elf64_Addr
+# endif
+#else
+  static inline Elf64_Addr
+#endif
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2)
+  __attribute ((always_inline))
+#endif
 elf_machine_plt_value (struct link_map *map, const Elf64_Rela *reloc,
 		       Elf64_Addr value)
 {
@@ -238,7 +304,14 @@
 /* Perform the relocation specified by RELOC and SYM (which is fully resolved).
    MAP is the object containing the reloc.  */
 
-static inline void
+#if __GNUC__ >= 4               
+  auto inline void
+#else
+  static inline void
+#endif
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2)
+  __attribute ((always_inline))
+#endif   
 elf_machine_rela (struct link_map *map, const Elf64_Rela *reloc,
 		  const Elf64_Sym *sym, const struct r_found_version *version,
 		  void *const reloc_addr_arg)
@@ -430,7 +503,14 @@
     }
 }
 
-static inline void
+#if __GNUC__ >= 4               
+  auto inline void
+#else
+  static inline void
+#endif
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2)
+  __attribute ((always_inline))
+#endif   
 elf_machine_rela_relative (Elf64_Addr l_addr, const Elf64_Rela *reloc,
 			   void *const reloc_addr_arg)
 {
@@ -438,7 +518,14 @@
   *reloc_addr = l_addr + reloc->r_addend;
 }
 
-static inline void
+#if __GNUC__ >= 4               
+  auto inline void
+#else
+  static inline void
+#endif
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2)
+  __attribute ((always_inline))
+#endif   
 elf_machine_lazy_rel (struct link_map *map,
 		      Elf64_Addr l_addr, const Elf64_Rela *reloc)
 {
@@ -476,7 +563,18 @@
 /* Set up the loaded object described by L so its unrelocated PLT
    entries will jump to the on-demand fixup code in dl-runtime.c.  */
 
-static inline int
+#if __GNUC__ >= 4
+# ifdef RESOLVE
+    auto inline int
+# else
+    static inline int
+# endif
+#else
+  static inline int
+#endif
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 2)
+  __attribute ((always_inline))
+#endif
 elf_machine_runtime_setup (struct link_map *l, int lazy, int profile)
 {
   if (l->l_info[DT_JMPREL] && lazy)
